module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -502,58 +507,197 @@ protected function escapeCql($in)
     }
 
     /**
-     * Retrieve FOLIO instance using VuFind's chosen bibliographic identifier.
+     * Get FOLIO records by batches of ids
      *
-     * @param string $bibId Bib-level id
+     * @param string[] $ids         ids to look for in the records
+     * @param string   $idField     field to compare to given ids
+     * @param string   $responseKey response key with the records to retrieve
+     * @param string   $endpoint    FOLIO API endpoint
+     * @param string   $querySuffix optional string to append to the queries
      *
-     * @return object
+     * @return \Generator<object>
+     * @throws ILSException if there is an issue with the FOLIO response
      */
-    protected function getInstanceByBibId($bibId)
+    protected function getByBatch($ids, $idField, $responseKey, $endpoint, $querySuffix = '')
+    {
+        if (count($ids) == 0) {
+            return;
+        }
+        $idToKey = fn ($id) => $endpoint . '[' . $idField . '=' . $id . ']';
+        $idsToLookFor = [];
+        foreach ($ids as $id) {
+            $items = $this->getCachedData($idToKey($id));
+            if ($items == null) {
+                $idsToLookFor[] = $id;
+            } else {
+                foreach ($items as $item) {
+                    yield $item;
+                }
+            }
+        }
+        $resultsToCache = [];
+        foreach (array_chunk($idsToLookFor, self::QUERY_BY_IDS_BATCH_SIZE) as $idsInBatch) {
+            $idsWithQuotes = array_map(fn ($id) => '"' . $this->escapeCql($id) . '"', $idsInBatch);
+            $query = [
+                'query' => $idField . ' == (' . implode(' OR ', $idsWithQuotes) . ')' . $querySuffix,
+            ];
+            foreach (
+                $this->getPagedResults(
+                    $responseKey,
+                    $endpoint,
+                    $query
+                ) as $item
+            ) {
+                $key = $idToKey($item->$idField);
+                if (isset($resultsToCache[$key])) {
+                    $resultsToCache[$key][] = $item;
+                } else {
+                    $resultsToCache[$key] = [ $item ];
```

Comment:
Minor style suggestion: we don't usually space brackets like this (though clearly the style checker doesn't mind, so feel free to ignore me if you prefer it the other way).

```suggestion
                    $resultsToCache[$key] = [$item];
```

Reply:
1) `getByBatch` does return records in an unpredictable order. This is by design, because a particular order would only make sense for some cases, and also for performance. When using an `idField` that is not unique (like `getHoldingsByInstanceIds`), it would not be very useful to sort by ids. Also, when using a unique `idField`, `getByBatch` does not check if all records were found. I updated the comments to clarify.
`getHoldings` returns results in the given order.

2) The only reason that code would not execute is if `getPagedResults` throws an exception, and in that case we would not want to store partial data in the cache. There is an array of items for each key, and we would not want to store an incomplete array for a given key. Maybe we could move writing to the cache into the outer foreach loop, but I would not expect this to make a difference in practice (if `getPagedResults` throws an exception, there is a bigger issue than performance).

Reply:
Thanks for the clarifications; sounds like this is fine as-is, and the improved comments should help prevent future issues!